### PR TITLE
ITEP-71155 - Update training job flow

### LINKED ISSUE #776: Adapt training job to send new configuration to trainer
Send the revamped configuration to OTX when initiating the training process.

As per the design, the format should be YAML, matching the model manifest.

Furthermore, after training, update the model configuration with the `advanced_configs` parameters, which will be used for display purposes only.

This task should be completed after the OTX implementation is done and a new version is available.

### LINKED ISSUE #815: Modify Geti OTX scripts to be able to run training with new model manifests and OTX 2.5
OTX v2.5.0 introduces API changes, including a new model manifest format.
These changes break compatibility with OTXConfig and ConfigConverter. To ensure that training and optimization tasks continue to run properly with the updated training backend and configuration format, we need to apply the necessary adjustments accordingly.

### LINKED ISSUE #877: Remove dependency to exportable code from /optimized_models export endpoint
The `/optimized_models/{model.optimized_model_id}/export` endpoint ([source](https://github.com/open-edge-platform/geti/blob/2c9530673a5e05a336058aba4e1c5d010d542f95/interactive_ai/services/resource/app/communication/rest_endpoints/model_endpoints.py#L151)) returns a zip package containing `model.xml`, `model.bin` and, depending on the value of parameter `model_only`, also a `config.json` file.

This endpoint is accessed by two places:
- the UI, with `model_only=True`, when downloading a model from the Models page
- the model registration service ([source](https://github.com/open-edge-platform/geti/blob/2c9530673a5e05a336058aba4e1c5d010d542f95/interactive_ai/services/model_registration/app/service/model_converter.py#L377))

In the current implementation, the endpoint retrieves these files from the "exportable code", a zip archive generated by OTX during training. From OTX 2.5, however, this artifact won't be available anymore, so the endpoint must be updated to obtain the files elsewhere:
- `model.bin` and `model.xml` are readily available through `model.model_adapters`.
- `config.json` can be crafted based exclusively on information that can be found within `model.xml`

### LINKED ISSUE #878: Remove dependency to exportable code from deployment endpoint
The deployment package endpoint (`deployment_package:download`) with `package_type=geti_sdk` depends on the "exportable code", that is an OTX training artifact:

https://github.com/open-edge-platform/geti/blob/8976e8209a98db083f554f0886aaacc7f640dd37/interactive_ai/services/resource/app/resource_management/deployment_package_manager.py#L93

Starting from OTX 2.5, the exportable code zip archive won't be created anymore. To ensure the correct functioning of the deployment package, the implementation must be updated to source the necessary model files elsewhere. The structure of the deployed zip should be:
```
deployment/
    <task_type>/
        model/
            config.json
            model.bin
            model.xml
        model.json
        python/
            requirements.txt
    project.json
example_code
    ...
LICENSE
README.md
sample_image.jpg
```
This is very similar to the current structure, with the following differences:
- `model.xml` and `model.bin` are obtained directly from the model (i.e. from the training artifacts)
- `config.json` is reconstructed based on `model.xml`
- The whole content of the `python` folder is removed, except for `requirements.txt` that's needed by the SDK ([source](https://github.com/open-edge-platform/geti-sdk/blob/01c682ee5f2bb7b54154f63307039202a758e8ca/geti_sdk/deployment/deployed_model.py#L175))

### LINKED ISSUE #906: Model deprecations for 2.12
The following models can be deprecated from v2.12:
- Classification
  - Custom_Image_Classification_EfficientNet-V2-L
  - Custom_Image_Classification_MobileNet-V3-small
- Detection
  - Object_Detection_ResNeXt101_ATSS
  - ~Custom_Object_Detection_Gen3_SSD~
  - ~Custom_Object_Detection_YOLOX (tiny)~
  - Object_Detection_RTDetr_18
  - Object_Detection_RTDetr_101
  - Object_Detection_RTMDet_tiny 
- Semantic segmentation
  - Custom_Semantic_Segmentation_Lite-HRNet-x-mod3_OCR 
- Instance segmentation
  - Custom_Counting_Instance_Segmentation_MaskRCNN_ResNet50 
- Rotated detection
  - Custom_Rotated_Detection_via_Instance_Segmentation_MaskRCNN_ResNet50 

Since some of these models were the recommended ones for accuracy/speed/balance, also apply the following adjustments to the recommendations:
- Detection
  - Balance -> Custom_Object_Detection_Gen3_ATSS
  - Accuracy -> Object_Detection_DFine_X
  - Speed -> Object_Detection_YOLOX_S
- Semantic segmentation
  - Accuracy -> Custom_Semantic_Segmentation_DINOV2_S
- Rotated detection
  - default + accuracy -> Custom_Rotated_Detection_MaskRCNN_ResNet50_v2
- Instance segmentation
  - balance -> Custom_Instance_Segmentation_MaskRCNN_ResNet50_v2

### PR DESCRIPTION
**Pull request by maxxgx**
## 📝 Description

JIRA: ITEP-71155

- *Geti -> OTX*: replace "config.json" with "config.yaml" containing yaml configuration, for example:
```yaml
job_type: train
model_manifest_id: Custom_Image_Classification_EfficinetNet-B0
hyperparameters:
  data_preparation: 
    augmentation:
      horizontal_flip: true,
      vertical_flip: false,
    ...
  training:
    max_epochs: 100,
    batch_size: 32,
    learning_rate: 0.0049
    ...
  evaluation:
    ...
export_models:
  - format: openvino
    output_model_id: 67a32c53d8c2d927812b0b3b
    precision: FP32
    with_xai: true
  - format: openvino
    output_model_id: 67a32c53d8c2d927812b0b3c
    precision: FP32
    with_xai: false
  - format: openvino
    output_model_id: 67a32c53d8c2d927812b0b3d
    precision: FP16
    with_xai: false
  - format: onnx
    output_model_id: 67a32c53d8c2d927812b0b3e
    precision: FP32
    with_xai: false
```
- *OTX -> Geti*: return `advanced_config.json` containing extra configuration used by OTX, which was NOT included in the input config.yaml. The “advanced” configuration parameters cannot be set by the user because they lack backend definitions and are therefore only used for visualization. Example:
```json
[
    {
        "key": "optimum_confidence_threshold",	 
        "name": "Optimum confidence threshold",
        "description": "The confidence threshold for ideal predictions",
        "value": 0.65
    }
]
```

Also resolves issues:
- https://github.com/open-edge-platform/geti/issues/877
- https://github.com/open-edge-platform/geti/issues/878

## ✨ Type of Change

Select the type of change your PR introduces:

- [ ] 🐞 **Bug fix** – Non-breaking change which fixes an issue
- [x] 🚀 **New feature** – Non-breaking change which adds functionality
- [ ] 🔨 **Refactor** – Non-breaking change which refactors the code base
- [ ] 💥 **Breaking change** – Changes that break existing functionality
- [ ] 📚 **Documentation update**
- [ ] 🔒 **Security update**
- [ ] 🧪 **Tests**

## 🧪 Testing Scenarios

Describe how the changes were tested and how reviewers can test them too:

- [x] ✅ Tested manually
- [x] 🤖 Run automated end-to-end tests

## ✅ Checklist

Before submitting the PR, ensure the following:

- [x] 🔍 PR title is clear and descriptive
- [x] 📝 For internal contributors: If applicable, include the JIRA ticket number (e.g., ITEP-123456) in the PR **title**. Do **not** include full URLs
- [x] 💬 I have commented my code, especially in hard-to-understand areas
- [ ] 📄 I have made corresponding changes to the documentation
- [x] ✅ I have added tests that prove my fix is effective or my feature works

### CODE REVIEW

File: interactive_ai/workflows/geti_domain/train/job/tasks/prepare_and_train/train_helpers.py [hunk 1]
```diff
@@ -193,7 +190,7 @@ def _get_export_parameters(
         }:
             export_parameters.append(
                 {
-                    "type": model.model_format.name.lower(),
+                    "format": model.model_format.name.lower(),
```
Comment (leoll2) [inline interactive_ai/workflows/geti_domain/train/job/tasks/prepare_and_train/train_helpers.py]: Shouldn't this change be behind FF?

File: interactive_ai/workflows/geti_domain/optimize/job/tasks/helpers.py [hunk 2]
```diff
@@ -45,17 +44,16 @@ def _prepare_s3_bucket(
...
         export_parameters=[
             {
-                "type": "openvino",
+                "format": "openvino",
                 "output_model_id": input_model.id_,
                 "precision": ModelPrecision.INT8.name,
                 "with_xai": False,
             },
         ],
-        optimization_type=optimization_cfg.model_to_optimize.optimization_type,
```
Comment (leoll2) [inline interactive_ai/workflows/geti_domain/optimize/job/tasks/helpers.py]: Is this optional?

File: interactive_ai/workflows/geti_domain/optimize/job/tasks/helpers.py [hunk 3]
```diff
@@ -45,17 +44,16 @@ def _prepare_s3_bucket(
     adapter.push_placeholders()
     adapter.push_metadata()
     adapter.push_input_configuration(
-        model_template_id=model_template.model_template_id,
+        model_manifest_id=input_model.model_storage.model_manifest_id,
         hyper_parameters=hyper_parameter_dict,
         export_parameters=[
             {
-                "type": "openvino",
```
Comment (leoll2) [inline interactive_ai/workflows/geti_domain/optimize/job/tasks/helpers.py]: Should be behind FF?

File: interactive_ai/workflows/geti_domain/train/job/tasks/prepare_and_train/train_helpers.py (hunk 1 above)
Comment (maxxgx) [reply interactive_ai/workflows/geti_domain/train/job/tasks/prepare_and_train/train_helpers.py]: No, I don't think that's possible because OTX would have to be aware of the FF in order to handle both the old `config.json` and the new `config.yaml`

File: interactive_ai/workflows/geti_domain/optimize/job/tasks/helpers.py (hunk 2 above)
Comment (maxxgx) [reply interactive_ai/workflows/geti_domain/optimize/job/tasks/helpers.py]: It was never used

File: interactive_ai/workflows/geti_domain/train/job/tasks/prepare_and_train/train_helpers.py (hunk 1 above)
Comment (leoll2) [reply interactive_ai/workflows/geti_domain/train/job/tasks/prepare_and_train/train_helpers.py]: But the current OTX version expects the field "type" instead of "format", no? If you merge the PR, then this part would break. Am I missing something?

Comment (maxxgx) [reply interactive_ai/workflows/geti_domain/train/job/tasks/prepare_and_train/train_helpers.py]: Yes, we can't merge it now. I think we need to wait for the updated OTX version, which should only support the new `config.yaml` format. Then, if the feature flag is off, we can convert the legacy hyperparameters to the new format so that we always pass the new format to OTX.

File: interactive_ai/workflows/train/pyproject.toml [hunk 4]
```diff
@@ -9,6 +9,7 @@ dependencies = [
     "geti-types~=1.0.2",
     "geti-k8s-tools==1.0.0",
     "workflows_common[evaluation, shard-dataset]",
+    "pyyaml~=6.0",
```
Comment (leoll2) [inline interactive_ai/workflows/train/pyproject.toml:12]: `workflows_common` already depends on `geti-configuration-tools`; is it then necessary to declare the `pyyaml` dependency here?

File: interactive_ai/workflows/common/jobs_common_extras/experiments/adapters/ml_artifacts.py [hunk 5]
```diff
@@ -415,19 +418,20 @@ def _update_base_model(self, model: Model, precision: str, xai_suffix: str) -> N
...
+    def pull_output_configuration(self) -> dict:
+        # advanced_config.json contains extra configuration used by OTX that were NOT included in the input config.yaml
         data = _check_bytes_type(
             self.binary_repo.get_by_filename(
                 filename=os.path.join(
                     self.dst_path_prefix,
                     "outputs",
                     "configurations",
-                    "optimized-config.json",
+                    "advanced_config.json",
```
Comment (leoll2) [inline interactive_ai/workflows/common/jobs_common_extras/experiments/adapters/ml_artifacts.py]: Why is advanced config a JSON, while config is YAML?

Comment (maxxgx) [reply interactive_ai/workflows/common/jobs_common_extras/experiments/adapters/ml_artifacts.py]: It should be a json file because the backend should not do any post processing on it and with a JSON file we can directly append it in the REST view

File: interactive_ai/workflows/train/pyproject.toml (hunk 4 above)
Comment (maxxgx) [reply interactive_ai/workflows/train/pyproject.toml:12]: I think it's better to have keep this dependency here since the `train` workflow does not directly use `geti-configuration-tools` but `pyyaml` is always required.

File: interactive_ai/services/resource/pyproject.toml [hunk 6]
```diff
@@ -18,7 +18,8 @@ dependencies = [
     "geti-types",
     "iai_core",
     "media_utils",
-    "grpc_interfaces"
+    "grpc_interfaces",
+    "defusedxml~=0.7.1",
```
Comment (leoll2) [inline interactive_ai/services/resource/pyproject.toml]: nitpicky: external deps are grouped at the top of the list

File: interactive_ai/workflows/train/trainer/pyproject.toml [hunk 7]
```diff
@@ -9,14 +9,15 @@ dependencies = [
...
-    "otx[base]==2.4.6",
+    "otx",
     "geti-kafka-tools",
     "geti-telemetry-tools[telemetry,kafka]",
 ]
 
 [tool.uv.sources]
 geti-kafka-tools = { path = "../../../../libs/kafka_tools" }
 geti-telemetry-tools = { path = "../../../../libs/telemetry_tools" }
+otx = { git = "https://github.com/open-edge-platform/training_extensions.git", branch = "test_get_2.12" }
```
Comment (leoll2) [inline interactive_ai/workflows/train/trainer/pyproject.toml]: TODO this must change before the final release to point to an official release instead of a dev branch

File: interactive_ai/services/resource/pyproject.toml (hunk 6 above)
Comment (maxxgx) [reply interactive_ai/services/resource/pyproject.toml]: Done

File: interactive_ai/services/auto_train/Dockerfile [hunk 8]
```diff
@@ -3,7 +3,8 @@ FROM python:3.10-slim-bookworm@sha256:9dd6774a1276178f94b0cc1fb1f0edd980825d0ea7
 FROM base AS build
 
 # Install Git and clone the OTX at the specified tag
-ARG OTX_RELEASE_TAG=2.4.6
+# TODO: Update OTX_RELEASE_TAG to 2.5 once it is released
+ARG OTX_RELEASE_TAG=test_get_2.12
```
Comment (leoll2) [inline interactive_ai/services/auto_train/Dockerfile]: Templates will be removed in OTX 2.5 (based on @kprokofi PR: https://github.com/open-edge-platform/training_extensions/pull/4477). So you have two choices here:
- Update the services Dockerfiles, remove the logic to copy the model templates from OTX - they are no longer needed
- Leave the services Dockerfiles unchanged for now, i.e. still copy them from OTX 2.4 (even if unused), and postpone the removal to the template copying logic to the following release.

File: interactive_ai/workflows/train/trainer/pyproject.toml (hunk 7 above)
Comment (kprokofi) [reply interactive_ai/workflows/train/trainer/pyproject.toml]: We can use now release/2.5 branch. "test.." branch will be deleted

File: interactive_ai/services/auto_train/Dockerfile (hunk 8 above)
Comment (kprokofi) [reply interactive_ai/services/auto_train/Dockerfile]: We agreed to keep it one more release

File: interactive_ai/libs/iai_core_py/iai_core/entities/task_node.py [hunk 9]
```diff
@@ -50,6 +51,11 @@ def from_model_template(cls, model_template: ModelTemplate) -> "TaskProperties":
             is_anomaly=model_template.task_type.is_anomaly,
         )
 
+    @property
+    def is_annotation_filtering_supported(self) -> bool:
```
Comment (leoll2) [inline interactive_ai/libs/iai_core_py/iai_core/entities/task_node.py]: Rename suggestion: `has_annotations_with_area`

**Review by leoll2 — DISMISSED** [review]

File: interactive_ai/libs/iai_core_py/iai_core/entities/task_node.py (hunk 9 above)
Comment (maxxgx) [reply interactive_ai/libs/iai_core_py/iai_core/entities/task_node.py]: Renamed

**Review by leoll2 — APPROVED** [review]

**Review by piotrgrubicki — APPROVED** [review]
